### PR TITLE
Improvements around torch.cat on empty Variables

### PR DESCRIPTION
**Pull request by gchanan**
1. Test forward with requires_grad=False (this wasn't currently broken but was reported recently).
1.  Test forward with requires_grad=True (this was actually fixed recently in ATen when making size/stride native functions)
2.  Fix and test backwards.  This required changing the gradient formula and also changing gradcheck to support empty inputs.

Gradgradcheck fails due to some size issues that don't look related to the formula for this specific function.  It may not be able to handle multiple outputs where some are empty tensors.

### CODE REVIEW

**Comment by soumith** [comment]
@gchanan can you rebase this?
